include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h

```diff
@@ -2949,6 +2957,94 @@ MGTwoLevelTransfer<dim, LinearAlgebra::distributed::Vector<Number>>::
 
 
 
+template <int dim, typename Number>
+void
+MGTwoLevelTransfer<dim, LinearAlgebra::distributed::Vector<Number>>::reinit(
+  const DoFHandler<dim> &          dof_handler_fine,
+  const DoFHandler<dim> &          dof_handler_coarse,
+  const AffineConstraints<Number> &constraint_fine,
+  const AffineConstraints<Number> &constraint_coarse,
+  const unsigned int               mg_level_fine,
+  const unsigned int               mg_level_coarse)
+{
+  // determine if polynomial transfer can be performed via the following two
+  // criteria:
+  // 1) multigrid levels can be only used with polynomial transfer
+  bool do_polynomial_transfer =
+    (mg_level_fine != numbers::invalid_unsigned_int) ||
+    (mg_level_coarse != numbers::invalid_unsigned_int);
+
+  // 2) the meshes are identical
+  if (do_polynomial_transfer == false)
+    {
+      const internal::CellIDTranslator<dim> cell_id_translator(
+        internal::n_coarse_cells(dof_handler_fine),
+        internal::n_global_levels(dof_handler_fine));
+
+      AssertDimension(internal::n_coarse_cells(dof_handler_fine),
+                      internal::n_coarse_cells(dof_handler_coarse));
+      AssertIndexRange(internal::n_global_levels(dof_handler_coarse),
+                       internal::n_global_levels(dof_handler_fine) + 1);
+
+      IndexSet is_locally_owned_fine(cell_id_translator.size());
+      IndexSet is_locally_owned_coarse(cell_id_translator.size());
+
+      for (const auto &cell : dof_handler_fine.active_cell_iterators())
+        if (!cell->is_artificial() && cell->is_locally_owned())
+          is_locally_owned_fine.add_index(cell_id_translator.translate(cell));
+
+      for (const auto &cell : dof_handler_coarse.active_cell_iterators())
+        if (!cell->is_artificial() && cell->is_locally_owned())
+          is_locally_owned_coarse.add_index(cell_id_translator.translate(cell));
```

Comment:
I'll remove all `IndexSet::add_index()` usages from this class in a follow-up PR.